### PR TITLE
fix: Test all generic args for trait when finding matching impl

### PR DESCRIPTION
**Pull request by lowr**
Addresses https://github.com/rust-lang/rust-analyzer/pull/13463#issuecomment-1287816680

When finding matching impl for a trait method, we've been testing the unifiability of self type. However, there can be multiple impl of a trait for the same type with different generic arguments for the trait. This patch takes it into account and tests the unifiability of all type arguments for the trait (the first being the self type) thus enables rust-analyzer to find the correct impl even in such cases.

### CODE REVIEW

**Comment by flodiebold** [comment]
I think you should be able to unify the substitutions directly instead of looping manually? That would also take care of consts.

**Comment by flodiebold** [comment]
Also, I think we have some bug report(s) that this fixes as well.

**Comment by lowr** [comment]
@flodiebold I was pretty sure there's a way to do that! Are you talking about `InferenceTable::try_unify()`, which takes `T: Zip` instead of `Ty`?

Could someone r-?

**Comment by Veykril** [comment]
bors r-
(wouldn't be a problem to do that in a follow up PR either)

File: crates/hir-ty/src/method_resolution.rs [hunk 1]
```diff
@@ -646,30 +651,31 @@ pub fn lookup_impl_method(
...
-                .unify(self_ty, &impl_ty)
-                .then(|| {
-                    let wh_goals =
-                        crate::chalk_db::convert_where_clauses(db, impl_.into(), &substs)
-                            .into_iter()
-                            .map(|b| b.cast(Interner));
-
-                    let goal = crate::Goal::all(Interner, wh_goals);
+            if !table
+                .unify(trait_ref.substitution.as_slice(Interner), expected_subst.as_slice(Interner))
```
Comment (flodiebold) [inline crates/hir-ty/src/method_resolution.rs:669]: maybe we should just have actual `TraitRef`s here, then we could just unify those

File: crates/hir-ty/src/method_resolution.rs [hunk 2]
```diff
@@ -625,17 +625,22 @@ pub(crate) fn iterate_method_candidates<T>(
 }
 
 pub fn lookup_impl_method(
-    self_ty: &Ty,
     db: &dyn HirDatabase,
     env: Arc<TraitEnvironment>,
     trait_: TraitId,
     name: &Name,
+    fn_subst: Substitution,
```
Comment (flodiebold) [inline crates/hir-ty/src/method_resolution.rs:632]: kind of weird to be passing the fn subst, but not the actual fn ID, right? we could just pass the trait subst, and then merge that with the `trait_` parameter into a `TraitRef`

Comment (lowr) [reply crates/hir-ty/src/method_resolution.rs:632]: Yeah I feel weird too. This function is called from hir and I didn't want to let hir manipulate `Substitution`s (besides passing them around as-is). It's just my preference so can we just make new `Substitution` out of fn subst in hir layer?

Comment (flodiebold) [reply crates/hir-ty/src/method_resolution.rs:632]: Hmm I think ideally we'd have some nice functions to get the trait ref from a function subst and so on. I think for now either would be fine.

**Comment by lowr** [comment]
So I ended up restructuring the related methods.

- `hir-ty::method_resolution::lookup_impl_method()` is now taking func id and func subst and is responsible for retrieving trait id and trait subst.
  - You can still pass any (unrelated) func id and func subst to the method, which I eventually want to forbid to prevent logical errors. One way to do this is to make an extension method for `TyKind::FnDef`, which ties up func id and subst, but we need to intern them in [some code paths](https://github.com/rust-lang/rust-analyzer/blob/53b6d69e931a52cca44ef94c3b713b4d255754b8/crates/hir/src/source_analyzer.rs#L271-L273) which would then immediately looked up, and that's... inefficient.
- `SourceAnalyzer::resolve_impl_method()` has been merged into `resolve_impl_method_or_trait_def()` because all the callers of those methods return the original func id when the lookup failed.

@flodiebold I'd appreciate if you could take a look again. I think the weirdness is somewhat alleviated and it's more organized than before.

**Pull request merged**